adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1209,24 +1214,34 @@ protected final void onCommandReceived(
         // example: control/DEFAULT_TENANT/4711/req/xyz/light
         final String topic = String.format("%s/%s/%s/%s/%s/%s", subscription.getEndpoint(), tenantId, deviceId,
                 subscription.getRequestPart(), command.getRequestId(), command.getName());
+        LOG.debug("Publishing command to device [tenant-id: {}, device-id: {}, MQTT client-id: {}, QoS: {}]",
+                subscription.getTenant(), subscription.getDeviceId(), endpoint.clientIdentifier(), qos);
         endpoint.publish(topic, command.getPayload(), qos, false, false);
+        endpoint.publishAcknowledgeHandler(
+                messageId -> afterCommandPublished(endpoint, qos, subscription, topic, commandContext));
```

Comment:
this is not how the *publishAcknowledgeHandler* works, is it? The handler is supposed to process the message ID that is being acknowledged by the peer. So I guess it needs to be passed in to the handler method as a parameter. This, however, requires that you keep track of the unacknowledged message IDs somewhere ...

Reply:
I got it. I will make use of the message ID. Currently there is no provision in vertx-mqtt library to get the message ID while publishing a message so that it can be matched with the message ID during acknowledgement. There is a method `lastMessageId()`, which can be used to retrieve the message ID of the last sent message. In some cases, its not sure which message ID will be return. 
A new method `publish(String topic, Buffer payload, MqttQoS qosLevel, boolean isDup, boolean isRetain, Handler<AsyncResult<Integer>> publishSentHandler)` is commited into vertx-mqtt master from [PR#86](https://github.com/vert-x3/vertx-mqtt/pull/86), which can be used. Should we wait for the release of next release of vertx-mqtt (3.6.0)? WDYT.

Reply:
I think that would be the easiest way. In the meantime we will live with QoS 0 for publishing commands to devices. I was actually surprised to discover that this method isn't in the MqttEndpoint interface already. It is in the MqttClient and I thought that it should be in the endpoint as well ...

Reply:
now that we have upgraded to vert.x 3.6, you may want to continue to work on this?

Reply:
I am already working on it :-)